### PR TITLE
CartCard

### LINKED ISSUE #79: Cart card

### PR DESCRIPTION
**Pull request by BrennyChao**
ADD
```

const image = "https://i.imgur.com/IO6e5a6.jpg";
    const title = "Some very long title and stuff";
    const currentStock = 9; 
    const checkedOutQuantity = 3;
    const isError = false;

    const closeChange = (event) => {

    };

```
and add 

`<CartCard image={image} title={title} currentStock={currentStock} closeChange={closeChange} checkedOutQuantity={checkedOutQuantity} isError={isError}/>`

under `<p>IEEEEEE</p>`

### CODE REVIEW

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import {
+    Card,
+    CardMedia,
+    Select,
+    FormControl,
+    MenuItem,
+    IconButton,
+    Paper,
+    InputLabel,
+} from "@material-ui/core";
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: Separate these into individual import statements

```javascript
import Card from "@material-ui/core/Card";
import CardMedia from "@material-ui/core/CardMedia";
etc.
````

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                    className={styles.CartPic}
+                    image={image}
+                    alt={title}
+                    component="img"
+                />
+            </Paper>
+            <Card className={styles.CartParts} elevation={0}>
+                <Typography variant="body2" className={styles.CartName}>
+                    {" "}
+                    {title}{" "}
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: Remove all these `{" "}`

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        items.push(
+            <MenuItem key={i} role="quantity" value={i.toString()}>
+                {i}
+            </MenuItem>
+        );
+    }
+    return items;
+};
+
+export const CartCard = ({
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js:58]: After reading the Material UI Card API, I think it's best to structure your code like this:

```javascript
<Card>
    <CardMedia />
    <CardContent>
          {put all the typography and form dropdown inside here}
    </CardContent>
    <CardActions>
           <IconButton>
                    <CloseIcon />
           </IconButton>
      </CardActions>
</Card>
```

So remove Paper and Grid as well

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                    </Select>
+                </FormControl>
+            </Card>
+            <IconButton
+                size="small"
+                className={styles.CartClose}
+                onClick={closeChange}
+                role="close"
+            >
+                <CloseIcon className={styles.CartClose} />
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: IconButton and CloseIcon have the same class which doesn't make sense. Make 2 separate classes.

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.module.scss [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        color: color(font);
+        padding-top: 11px;
+        padding-bottom: 17px;
+    }
+
+    &Parts {
+        display: flex;
+        flex-direction: column;
+        height: 90px;
+        width: 95%;
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.module.scss]: Set width to 100%. On the Cart.js page, we'll set the width of the container the CartCard container is

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    return items;
+};
+
+export const CartCard = ({
+    image,
+    title,
+    currentStock,
+    closeChange,
+    handleChange,
+    item,
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: Another prop you need is the quantity of the object they checked out. The dropdown should be displaying that number

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                    variant="outlined"
+                    size="small"
+                    className={styles.CartFormControl}
+                >
+                    <InputLabel shrink={true} id="Qty">
+                        {" "}
+                        Qty{" "}
+                    </InputLabel>
+                    <Select
+                        labelId="Qty"
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: It should be `label` not `labelId`

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+} from "@material-ui/core";
+import CloseIcon from "@material-ui/icons/Close";
+
+const Selections = (currentStock) => {
+    const items = [];
+    let i = 0;
+
+    for (i = 1; i < currentStock + 1; i++) {
+        items.push(
+            <MenuItem key={i} role="quantity" value={i.toString()}>
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: You can just set value to `{i}`

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            <Paper className={styles.Cart} elevation={0}>
+                <CardMedia
+                    className={styles.CartPic}
+                    image={image}
+                    alt={title}
+                    component="img"
+                />
+            </Paper>
+            <Card className={styles.CartParts} elevation={0}>
+                <Typography variant="body2" className={styles.CartName}>
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js:80]: If you look on the Figma frame 7.1 - Cart Error, if there's an error, there's a message in red under the title that shows up. Have a prop called `isError` for now to handle it

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js (hunk 6 above)
Comment (lisa-sa-li) [reply hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: For now, you should move 

```javascript
const [item, setItem] = React.useState(checkedOutQuantity);
const handleChange = (event) => {
        setItem(event.target.value);
};
```
inside CartCard.js, and `checkedOutQuantity` should be an integer prop you pass in that is the quantity that they checked out

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
+import React from "react";
+import CartCard from "./CartCard";
+import { render } from "@testing-library/react";
+import "@testing-library/jest-dom/extend-expect";
+import { fireEvent } from "@testing-library/react";
+import { TEST_IDS } from "components/user/LoginForm/LoginForm";
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js]: remove this TEST_IDS line

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+describe("CartCard", () => {
+    const image = "https://i.imgur.com/IO6e5a6.jpg";
+    const title = "Some very long title and stuff";
+    const currentStock = 9;
+    const closeChange = jest.fn();
+    const handleChange = jest.fn();
+    const item = 3;
+
+    test("CartCard Test", () => {
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js]: You should split them into 2 separate tests. One that clicks the button, and the other to check that everything expected appears in the component.

**Review by lisa-sa-li — DISMISSED** [review]

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            <MenuItem key={i} role="quantity" value={i}>
+                {i}
+            </MenuItem>
+        );
+    }
+    return items;
+};
+
+const Stock = ({ currentStock, handleChange, item }) => {
+    if (currentStock === 0) {
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: `if (!currentStock) {`

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        return (
+            <Typography variant="caption" className={styles.CartError}>
+                {" "}
+                Currently unavailable{" "}
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: Remove `{" "}`

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+                alt={title}
+                component="img"
+            />
+            <CardContent className={styles.CartParts} elevation={0}>
+                <Typography variant="body2" className={nameStyle}>
+                    {title}
+                </Typography>
+                <Typography variant="caption" className={styles.CartError}>
+                    {errorMessage}
+                </Typography>
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: You should conditionally render the error message rather than rendering a blank line

i.e.

```javascript
{isError &&
   <Typography variant="caption" className={styles.CartError}>
       errorMessage}
    </Typography>
}
```

Comment (lisa-sa-li) [reply hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: After doing this, you can just set a message like `const errorMessage = "Some error message"` on line 73

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.module.scss [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    &Action {
+        background-color: color(backgroundC);
+        display: inline;
+    }
+
+    &Close {
+        height: 24px;
+        width: 24px;
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.module.scss]: You can remove the height and width bc the icon is default 24px

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.module.scss [hunk 16]
```diff
@@ -0,0 +1,53 @@
+@import "./../../../assets/abstracts/mixins";
+@import "./../../../assets/abstracts/variables";
+
+.Cart {
+    display: flex;
+    color: color(backgroundC);
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.module.scss]: This should be `background-color: color(backgroundC);`. You can remove the other `background-color: color(backgroundC);` classes below

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.module.scss [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+        color: color(secondary);
+        padding-top: 0px;
+        padding-bottom: 1px;
+    }
+
+    &Parts {
+        display: flex;
+        flex-direction: column;
+        height: 90px;
+        width: 100%;
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.module.scss]: Remove setting the height to a certain pixel (just remove the height attribute completely for now)

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.module.scss [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+        padding-left: 30px;
+        background-color: color(backgroundC);
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.module.scss]: Remove the background color

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+    checkedOutQuantity,
+    isError,
+}) => {
+    const [item, setItem] = React.useState(checkedOutQuantity);
+    const handleChange = (event) => {
+        setItem(event.target.value);
+    };
+    const errorMessage = isError === true ? "some error message" : "";
+    var nameStyle = isError === true ? styles.CartErrorName : styles.CartName;
+    nameStyle = currentStock === 0 ? styles.CartErrorName : nameStyle;
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: I understand you're trying to manipulate the padding here. I was playing around with the padding. If you just give `&FormControl` a `margin-top: 10px;` and remove padding from `&Name` and `&Error`, it works out. If you do this, there's no more need for `&ErrorName`

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js [hunk 20]
```diff
@@ -0,0 +1,33 @@
+import React from "react";
```
Comment (lisa-sa-li) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js:1]: Add more tests to test the error functionality. 

i.e. if currentStock is 0, expect "Currently unavailable" to be in document.
if isError = true, expect "Some error message" to be in document.
Change the dropdown, expect whatever number to be in document

Comment (BrennyChao) [reply hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js:1]: I think I added this in "added the test, still missing line 70" commit that I made right after this if you want to take a look at that

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+import InputLabel from "@material-ui/core/InputLabel";
+import MenuItem from "@material-ui/core/MenuItem";
+import IconButton from "@material-ui/core/IconButton";
+import CloseIcon from "@material-ui/icons/Close";
+
+const Selections = (currentStock) => {
+    const items = [];
+    let i = 0;
+
+    for (i = 1; i < currentStock + 1; i++) {
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: You can do the `let i=0` inline here. For clarity, I would make this

```javascript
for (let i=1; i <= currentStock; i++) {
   ...
}
```

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+                className={styles.CartFormControl}
+            >
+                <InputLabel shrink={true} id="Qty">
+                    Qty
+                </InputLabel>
+                <Select
+                    label="Qty"
+                    value={item}
+                    onChange={handleChange}
+                    role="selecter"
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: Do you actually need this role? Seems the only `getByRole` that uses this is commented out.

Also, `selector` is the proper spelling.

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+
+export const CartCard = ({
+    image,
+    title,
+    currentStock,
+    closeChange,
+    checkedOutQuantity,
+    isError,
+}) => {
+    const [item, setItem] = React.useState(checkedOutQuantity);
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: This is fine for now, but eventually this will have to be replaced with a Formik wrapper that handles the form state. Do you want to do that in a separate PR? You can take a look at what that would involve in the `ProductOverview` component: https://github.com/ieeeuoft/hackathon-template/blob/develop/hackathon_site/dashboard/frontend/src/components/inventory/ProductOverview/ProductOverview.js#L80

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+    const handleChange = (event) => {
+        setItem(event.target.value);
+    };
+
+    const errorMessage = "some error message";
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: Instead of this, can you make a prop that's just `error`, a string. If it's not null, then show it.

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+                className={styles.CartPic}
+                image={image}
+                alt={title}
+                component="img"
+            />
+            <CardContent className={styles.CartParts} elevation={0}>
+                <Typography variant="body2" className={styles.CartName}>
+                    {title}
+                </Typography>
+                {isError && (
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: With the change to an `error` prop,

```javascript
{error && (
    <Typography variant="caption" className={styles.CartError}>
        {error}
    </Typography>
)}
```

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+        items.push(
+            <MenuItem key={i} role="quantity" value={i}>
+                {i}
+            </MenuItem>
+        );
+    }
+    return items;
+};
+
+const Stock = ({ currentStock, handleChange, item }) => {
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: `item` isn't the best prop name. Can you use something more descriptive of what it is?

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+                image={image}
+                title={title}
+                currentStock={currentStock}
+                closeChange={closeChange}
+                checkedOutQuantity={checkedOutQuantity}
+                isError={isError}
+            />
+        );
+
+        // const selectedValue = getByRole("selecter");
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js]: Remove

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+        const selectedValue = screen.getByDisplayValue(checkedOutQuantity.toString());
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js]: Did you need to use `screen`? It's fine if you need to, but for selecting dropdown elements and such `getByText` will work. See the `ProductOverview` tests: https://github.com/ieeeuoft/hackathon-template/blob/develop/hackathon_site/dashboard/frontend/src/components/inventory/ProductOverview/ProductOverview.test.js

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+    const image = "https://i.imgur.com/IO6e5a6.jpg";
+    const title = "Some very long title and stuff";
+    const currentStock = 9;
+    const closeChange = jest.fn();
+    const checkedOutQuantity = 3;
+    const isError = false;
+    const outOfStock = "Currently unavailable";
+    const errorMessage = "some error message";
+
+    test("CartCard Test", () => {
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js]: More descriptive string about what's being tested

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+            </FormControl>
+        );
+    }
+};
+
+export const CartCard = ({
+    image,
+    title,
+    currentStock,
+    closeChange,
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: I would make this `handleRemove`, since the button is to remove an item specifically (not just close the card)

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js [hunk 31]
```diff
@@ -0,0 +1,88 @@
...
+                currentStock={currentStock}
+                closeChange={closeChange}
+                checkedOutQuantity={checkedOutQuantity}
+                isError={isError}
+            />
+        );
+
+        const exitButton = getByRole("close");
+        fireEvent.click(exitButton);
+        expect(closeChange.mock.calls.length).toBe(1);
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js]:
```javascript
expect(closeChange).toHaveBeenCalledTimes(1)`
```

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js [hunk 32]
```diff
@@ -0,0 +1,88 @@
...
+                currentStock={currentStock}
+                closeChange={closeChange}
+                checkedOutQuantity={checkedOutQuantity}
+                isError={isError}
+            />
+        );
+
+        // const selectedValue = getByRole("selecter");
+        const selectedValue = screen.getByDisplayValue(checkedOutQuantity.toString());
+        // expect(selectedValue.value).toBe(checkedOutQuantity);
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js]: Remove if you don't need this

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js (hunk 23 above)
Comment (lisa-sa-li) [reply hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.js]: Lets do this in a separate PR

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+import CartCard from "./CartCard";
+import { render } from "@testing-library/react";
+import "@testing-library/jest-dom/extend-expect";
+import { fireEvent } from "@testing-library/react";
+
+describe("CartCard Item test", () => {
+    const image = "https://i.imgur.com/IO6e5a6.jpg";
+    const title = "Some very long title and stuff";
+    const currentStock = 9;
+    const handleRemove = jest.fn();
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js]: You have to be careful about defining spys outside of the particular test method, because the mock object will never be reset as-is. When you need to have a mock or spy, it's best to define it directly in the test that uses it. For the rest of the tests, you could leave this as a dumb function like `() => {}`.

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js [hunk 34]
```diff
@@ -0,0 +1,83 @@
...
+                error={error}
+            />
+        );
+
+        expect(getByText(checkedOutQuantity.toString())).toBeInTheDocument();
+        expect(getByAltText(title)).toBeInTheDocument();
+        expect(getByText(title)).toBeInTheDocument();
+    });
+
+    test("Button that removes the CartCard item", () => {
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js:33]: To reiterate above, inside this function is where you should be defining your spy/mock for `handleRemove`

File: hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js [hunk 35]
```diff
@@ -0,0 +1,85 @@
...
+        const handleRemove = jest.fn();
```
Comment (grahamhoyes) [inline hackathon_site/dashboard/frontend/src/components/cart/CartCard/CartCard.test.js:34]: This works, but just be wary that you're playing with variable scope here since you already have `const handleRemove` defined outside (this applies for all of the tests where you've overriding something).

For future reference, you can make a function that renders the component with some default props, and then also accepts new ones to override them. Something like

```javascript
const renderComponent = (customProps) => {
    return render(<CartCard image={defaultImage} title="Some default title" {...customProps} />);
};
```

`customProps` needs to be an object, so you would wind up in your test doing something like:

```javascript
const { getByRole } = renderComponent({handleRemove: handleRemoveSpy});
```

**Review by grahamhoyes — APPROVED** [review]
Tests look good, so feel free to squash and merge. In the future, be careful about redefining const variables in a lower scope, as you may run into some confusion. I've included a snippet for one common way to handle tests that just change one prop of the component, for you to consider in the future.

**Review by grahamhoyes — APPROVED** [review]
